meter/tapo/connection.go

```diff
@@ -52,365 +31,72 @@ type Connection struct {
-	if uri == "" {
-		return nil, errors.New("missing uri")
-	}
-
-	if user == "" || password == "" {
-		return nil, fmt.Errorf("missing user or password")
-	}
-
-	for _, suffix := range []string{"/", "/app"} {
-		uri = strings.TrimSuffix(uri, suffix)
-	}
-
-	log := util.NewLogger("tapo")
-
-	// nosemgrep:go.lang.security.audit.crypto.use_of_weak_crypto.use-of-sha1
-	h := sha1.New()
-	_, err := h.Write([]byte(user))
-	userhash := hex.EncodeToString(h.Sum(nil))
-
-	conn := &Connection{
-		log:             log,
-		Helper:          request.NewHelper(log),
-		URI:             fmt.Sprintf("%s/app", util.DefaultScheme(uri, "http")),
-		EncodedUser:     base64.StdEncoding.EncodeToString([]byte(userhash)),
-		EncodedPassword: base64.StdEncoding.EncodeToString([]byte(password)),
-	}
-
-	conn.Client.Timeout = Timeout
-
-	return conn, err
-}
-
-// Login provides the Tapo device session token and MAC address (TerminalUUID).
-func (d *Connection) Login() error {
-	err := d.Handshake()
-	if err != nil {
-		return err
-	}
-
-	req := map[string]interface{}{
-		"method": "login_device",
-		"params": map[string]interface{}{
-			"username": d.EncodedUser,
-			"password": d.EncodedPassword,
-		},
-	}
-
-	res, err := d.DoSecureRequest(d.URI, req)
-	if err != nil {
-		return err
-	}
-
-	if err := d.CheckErrorCode(res.ErrorCode); err != nil {
-		return err
-	}
-
-	d.Token = res.Result.Token
-
-	deviceResponse, err := d.ExecMethod("get_device_info", false)
-	if err != nil {
-		return err
-	}
-
-	d.TerminalUUID = deviceResponse.Result.MAC
-
-	return nil
-}
-
-// Handshake provides the Tapo device session cookie and encryption cipher.
-func (d *Connection) Handshake() error {
-	privKey, pubKey, err := GenerateRSAKeys()
-	if err != nil {
-		return err
-	}
-
-	pubPEM, err := DumpRSAPEM(pubKey)
+	url, err := url.Parse(uri)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("invalid url: %s", uri)
 	}
 
-	req, err := json.Marshal(map[string]interface{}{
-		"method": "handshake",
-		"params": map[string]interface{}{
-			"key":             string(pubPEM),
-			"requestTimeMils": 0,
-		},
-	})
+	addr, err := netip.ParseAddr(url.Hostname())
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("invalid ip address: %s", uri)
 	}
 
-	resp, err := http.Post(d.URI, "application/json", bytes.NewBuffer(req))
-	if err != nil {
-		return err
-	}
-	defer resp.Body.Close()
-
-	var res DeviceResponse
-	if err = json.NewDecoder(resp.Body).Decode(&res); err != nil {
-		return err
-	}
-
-	if err = d.CheckErrorCode(res.ErrorCode); err != nil {
-		return err
-	}
-
-	encryptedEncryptionKey, err := base64.StdEncoding.DecodeString(res.Result.Key)
-	if err != nil {
-		return err
+	if user == "" || password == "" {
+		return nil, fmt.Errorf("missing user or password")
 	}
 
-	encryptionKey, err := rsa.DecryptPKCS1v15(rand.Reader, privKey, encryptedEncryptionKey)
-	if err != nil {
-		return err
-	}
+	// log := util.NewLogger("tapo")
+	log := log.New(io.Discard, "", 0)
```

Comment:
@andig + @insomniacslk could U pls give an advice on how to deal with the different logging approaches? From my perspective it would be nice to enable at least a trace logging, in case of future user issues ...

Reply:
Seems doing so we lose all logging. It would be nice if tapo library allowed to log wire traffic, nothing else.